File: solutions/javascript/anagram/1/anagram.js
/** @param {string} str */
const normalize = (str) => str.split('').sort().join('');

/**
 * @param {string} word
 * @param {string[]} candidates
 */
export const findAnagrams = (word, candidates) => {
    const result = [];
    const lowerWord = word.toLowerCase();
    const normalizedWord = normalize(lowerWord);
    for (const candidate of candidates) {
        const lowerCandidate = candidate.toLowerCase();
        if (lowerCandidate === lowerWord) {
            continue;
        }
        const normalizedCandidate = normalize(lowerCandidate);
        if (normalizedWord === normalizedCandidate) {
            result.push(candidate);
        }
    }
    return result;
};

File: solutions/go/anagram/1/anagram.go
package anagram

import (
	"sort"
	"strings"
)

func normalize(s string) string {
	runes := []rune(s)
	sort.Slice(runes, func(i, j int) bool {
		return runes[i] < runes[j]
	})
	return string(runes)
}

func Detect(subject string, candidates []string) []string {
	result := []string{}
	lowerSubject := strings.ToLower(subject)
	normalizedSubject := normalize(lowerSubject)

	for _, candidate := range candidates {
		if lowerSubject == strings.ToLower(candidate) {
			continue
		}
		if normalizedSubject == normalize(strings.ToLower(candidate)) {
			result = append(result, candidate)
		}
	}

	return result
}
